README.md
# NAME

POE::Component::IRC::Plugin::WWW::Vim::Tips - IRC plugin to fetch Vim tips



# SYNOPSIS

    use strict;
    use warnings;

    use POE qw(Component::IRC  Component::IRC::Plugin::WWW::Vim::Tips);

    my $irc = POE::Component::IRC->spawn(
        nick    => 'nickname',
        server  => 'irc.freenode.net',
        port    => 6667,
        ircname => 'ircname',
    );

    POE::Session->create(package_states => [main => [qw(_start irc_001)]]);

    $poe_kernel->run;

    sub _start {
        $irc->yield(register => 'all');

        $irc->plugin_add('Vim-Tips' => POE::Component::IRC::Plugin::WWW::Vim::Tips->new);

        $irc->yield(connect => {});
    }

    sub irc_001 {
        $irc->yield(join => '#channel');
    }

# DESCRIPTION

type !vimtip or !vimtips to get a random Vim tip, currenly fetched from [http://twitter.com/vimtips](http://twitter.com/vimtips)

here is a cool one-liner if you just want the Vim tip:

    perl -Ilib -MPOE::Component::IRC::Plugin::WWW::Vim::Tips
         -le 'print POE::Component::IRC::Plugin::WWW::Vim::Tips->new->_get_vim_tip'

# AUTHOR

Curtis Brandt, `<curtis at cpan.org>`

# BUGS

Please report any bugs or feature requests to `bug-poe-component-irc-plugin-vimtips at rt.cpan.org`, or through
the web interface at [http://rt.cpan.org/NoAuth/ReportBug.html?Queue=POE-Component-IRC-Plugin-WWW-Vim-Tips](http://rt.cpan.org/NoAuth/ReportBug.html?Queue=POE-Component-IRC-Plugin-WWW-Vim-Tips).  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.







# SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc POE::Component::IRC::Plugin::WWW::Vim::Tips



You can also look for information at:

- RT: CPAN's request tracker (report bugs here)

    [http://rt.cpan.org/NoAuth/Bugs.html?Dist=POE-Component-IRC-Plugin-WWW-Vim-Tips](http://rt.cpan.org/NoAuth/Bugs.html?Dist=POE-Component-IRC-Plugin-WWW-Vim-Tips)

- AnnoCPAN: Annotated CPAN documentation

    [http://annocpan.org/dist/POE-Component-IRC-Plugin-WWW-Vim-Tips](http://annocpan.org/dist/POE-Component-IRC-Plugin-WWW-Vim-Tips)

- CPAN Ratings

    [http://cpanratings.perl.org/d/POE-Component-IRC-Plugin-WWW-Vim-Tips](http://cpanratings.perl.org/d/POE-Component-IRC-Plugin-WWW-Vim-Tips)

- Search CPAN

    [http://search.cpan.org/dist/POE-Component-IRC-Plugin-WWW-Vim-Tips/](http://search.cpan.org/dist/POE-Component-IRC-Plugin-WWW-Vim-Tips/)



# ACKNOWLEDGEMENTS

Thanks to Graham Barr for guidance in my Perl work



# LICENSE AND COPYRIGHT

Copyright 2013 Curtis Brandt.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


Changes
{{$NEXT}}

0.14  2014-01-11 11:17:24 PST
        twitter JSON gone, extract tweets with HTML instead

0.13  2013-06-24 10:35:08 PDT
        remove vim tip check from test
        add one liner to POD

0.12  2013-06-17 14:41:23 PDT
        add vim.t test

0.11  2013-06-10 11:03:36 PDT
        replace tests with basic.t. TODO: write tests

0.10  2013-06-09 00:04:12 PDT
        Update 00-load.t to use all prereq modules...will it help?

0.09  2013-06-08 21:06:05 PDT
        use Perl 5.8, not 5.6 / remove my name from contributers

0.08  2013-06-01 10:57:54 PDT
        Update external links and remove version from POD

0.07  2013-05-28 07:30:36 PDT
        Rename module as POE::Component::IRC::Plugin::WWW::Vim::Tips

0.06    2013-04-25
        Forgot to bump version in POD

0.05    2013-02-07
        Ack, I left a backup folder that Dist::Zilla found

0.04    2013-02-07
        Convert to Dist::Zilla
        Remove debug content

0.03    2013-02-02
        Switch from HTML::TreeBuilder::XPath to Mojo::JSON

0.02    2013-01-13
        Don't forget to list dependencies in Makefile.PL

0.01    2013-28-01
        First version




lib/POE/Component/IRC/Plugin/WWW/Vim/Tips.pm
package POE::Component::IRC::Plugin::WWW::Vim::Tips;

use 5.008_005;
use strict;
use warnings;

our $VERSION = '0.13';

use POE::Component::IRC::Plugin qw( :ALL );
use HTML::TreeBuilder::XPath;
use LWP::Simple qw(get);

sub new {
    my $package = shift;
    return bless {}, $package;
}

sub PCI_register {
    my ($self, $irc) = splice @_, 0, 2;

    $irc->plugin_register($self, 'SERVER', qw(public));
    return 1;
}

# This is method is mandatory but we don't actually have anything to do.
sub PCI_unregister {
    return 1;
}

sub S_public {
    my ($self, $irc) = splice @_, 0, 2;

    # Parameters are passed as scalar-refs including arrayrefs.
    my $nick    = (split /!/, ${$_[0]})[0];
    my $channel = ${$_[1]}->[0];
    my $msg     = ${$_[2]};

    my $reply =
        ($msg =~ /^!vimtips?\b/i)   ? $self->_get_vim_tip
      : ($msg =~ /^!emacstips?\b/i) ? 'use Vim'
      :                               '';
    if ($reply) {
        $irc->yield(privmsg => $channel => "$nick: $reply");
        return PCI_EAT_PLUGIN;
    }

    # Default action is to allow other plugins to process it.
    return PCI_EAT_NONE;
}

sub _get_vim_tip {
    my $content = get('http://twitter.com/vimtips');
    my $tree    = HTML::TreeBuilder::XPath->new_from_content($content);
    my @tips    = grep { s/\x{a0}/ /g; s/^\s*|\s*$//g; 1 }                #
      $tree->findvalues(
        './/div[@class=~/\bprofile-stream\b/]
          //p[@class=~/\btweet-text\b/]'
      );
    return $tips[rand @tips];
}

1;

__END__

=pod

=head1 NAME

POE::Component::IRC::Plugin::WWW::Vim::Tips - IRC plugin to fetch Vim tips


=head1 SYNOPSIS

    use strict;
    use warnings;

    use POE qw(Component::IRC  Component::IRC::Plugin::WWW::Vim::Tips);

    my $irc = POE::Component::IRC->spawn(
        nick    => 'nickname',
        server  => 'irc.freenode.net',
        port    => 6667,
        ircname => 'ircname',
    );

    POE::Session->create(package_states => [main => [qw(_start irc_001)]]);

    $poe_kernel->run;

    sub _start {
        $irc->yield(register => 'all');

        $irc->plugin_add('Vim-Tips' => POE::Component::IRC::Plugin::WWW::Vim::Tips->new);

        $irc->yield(connect => {});
    }

    sub irc_001 {
        $irc->yield(join => '#channel');
    }

=head1 DESCRIPTION

type !vimtip or !vimtips to get a random Vim tip, currenly fetched from L<http://twitter.com/vimtips>

here is a cool one-liner if you just want the Vim tip:

    perl -Ilib -MPOE::Component::IRC::Plugin::WWW::Vim::Tips
         -le 'print POE::Component::IRC::Plugin::WWW::Vim::Tips->new->_get_vim_tip'

=head1 AUTHOR

Curtis Brandt, C<< <curtis at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-poe-component-irc-plugin-vimtips at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=POE-Component-IRC-Plugin-WWW-Vim-Tips>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc POE::Component::IRC::Plugin::WWW::Vim::Tips


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=POE-Component-IRC-Plugin-WWW-Vim-Tips>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/POE-Component-IRC-Plugin-WWW-Vim-Tips>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/POE-Component-IRC-Plugin-WWW-Vim-Tips>

=item * Search CPAN

L<http://search.cpan.org/dist/POE-Component-IRC-Plugin-WWW-Vim-Tips/>

=back


=head1 ACKNOWLEDGEMENTS

Thanks to Graham Barr for guidance in my Perl work


=head1 LICENSE AND COPYRIGHT

Copyright 2013 Curtis Brandt.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;
